### ISSUE TITLE
Add Structured Markup to Pages

### ISSUE THREAD

**Issue by prikhi** [srid/zulip-archive#11]
This would help display posts as rich results in search result pages: 
https://developers.google.com/search/docs/guides/intro-structured-data
https://schema.org/docs/schemas.html

The few relevant types I can think of:

* Breadcrumbs: https://developers.google.com/search/docs/data-types/breadcrumb 
  https://schema.org/BreadcrumbList
* Article: https://developers.google.com/search/docs/data-types/article 
  https://schema.org/Article
* Or the DiscussionForumPosting Article-subtype: https://schema.org/DiscussionForumPosting
* Organization(for site name, logo, url to this repo, etc): https://schema.org/Organization

**Comment by srid** [comment]
There is also [The Open Graph protocol](https://ogp.me/) that is involved in "link previews" generated by the likes of Twitter, Slack, etc. I already use this for my website.

**Comment by prikhi** [comment]
Yup, the spec is here: https://ogp.me/ along with some validators at the bottom of the page.

Probably the most important to add for topics are:

* `title` - the name of the topic
* `image` - avatar of the topic creator?
* `url` - canonical url of the topic
* `description` - first 2-3 sentences of the first message in the topic?
* `site_name` - `Functional Programming Zulip Chat Archive`
* `type` - `article`?
* `article:published_time` - datetime of the first message in the topic
* `article:article:modified_time` - datetime of the last message in the topic
* `article:section` - the stream name the topic is in

**Comment by prikhi** [comment]
I might do the breadcrumbs soon. Do you have a preference of RDFa or Microdata? I usually do microdata but both are easy to implement.

Do you think I should move the SEO stuff like structured attributes & OGP stuff to another module like `SEO.hs` or just keep it all in `Main`?

**Comment by srid** [comment]
> I might do the breadcrumbs soon. 

Nice!

> Do you have a preference of RDFa or Microdata? I usually do microdata but both are easy to implement.

I never heard of them before, and as such have no preference either way. :-) [In Wikipedia](https://en.wikipedia.org/wiki/RDFa#RDFa_Lite), it seems to indicate that _RDFa Lite_ is an "unique and definitive standard alternative to Microdata" - but I don't know the primacy of that in the larger context.

> Do you think I should move the SEO stuff like structured attributes & OGP stuff to another module like SEO.hs or just keep it all in Main?

I guess separate module as long as things are DRY.

You might also want to consider creating data types ([example](https://github.com/srid/website/blob/a0e8db4/src/Website/View/OpenGraph.hs#L11-L19)) to hold the SEO data, so that the construction of them ([example](https://github.com/srid/website/blob/master/src/Website/Content.hs#L135-L166)) is decoupled from their rendering ([example](https://github.com/srid/website/blob/a0e8db4/src/Website/View/OpenGraph.hs#L23-L36)). This way you can put the data type and the rendering function in SEO.hs, but keep the function that constructs them close to Main.hs.
